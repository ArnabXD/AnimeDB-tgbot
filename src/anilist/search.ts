import { client, gql } from "./gqlClient.ts";

export interface SearchResponse {
  Page: {
    pageInfo: {
      total: number;
      currentPage: number;
      lastPage: number;
      hasNextPage: boolean;
      perPage: number;
    };
    media: {
      id: number;
      title: {
        romaji: string;
        english: string | null;
      };
    }[];
  };
}

export default async function search(
  key: string,
  type: "ANIME" | "MANGA" = "ANIME",
  page = 1,
) {
  const query = gql`
    query ($id: Int, $page: Int, $perPage: Int, $search: String) {
      Page (page: $page, perPage: $perPage) {
        pageInfo {
          total
          currentPage
          lastPage
          hasNextPage
          perPage
        }
        media (id: $id, search: $search, type:${type}) {
          id
          title {
            romaji
            english
          }
        }
      }
    }`;

  try {
    const data = await client.request<SearchResponse>(query, {
      search: key,
      page: page,
      perPage: 10,
    });
    return data.Page.pageInfo.total === 0 ? null : data;
  } catch (err) {
    console.error(err);
    return null;
  }
}

await search("naruto", "ANIME");
