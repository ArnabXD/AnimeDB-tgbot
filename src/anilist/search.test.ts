import { assertSnapshot } from "testing/snapshot.ts";
import search from "./search.ts";

Deno.test("search anime", async (snap) => {
  const result = await search("Boku no Pico", "ANIME");
  assertSnapshot(snap, result);
});

Deno.test("search manga", async (snap) => {
  const result = await search("Tokyo Revengers", "MANGA");
  assertSnapshot(snap, result);
});
