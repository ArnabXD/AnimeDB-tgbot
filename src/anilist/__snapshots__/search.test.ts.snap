export const snapshot = {};

snapshot[`search anime 1`] = `
{
  Page: {
    media: [
      {
        id: 1639,
        title: {
          english: null,
          romaji: "Boku no Pico",
        },
      },
      {
        id: 5391,
        title: {
          english: null,
          romaji: "Pico to Chico",
        },
      },
      {
        id: 4866,
        title: {
          english: null,
          romaji: "Pico x CoCo x Chico",
        },
      },
      {
        id: 6546,
        title: {
          english: null,
          romaji: "pico: Boku no Chiisana Natsu Monogatari",
        },
      },
    ],
    pageInfo: {
      currentPage: 1,
      hasNextPage: false,
      lastPage: 1,
      perPage: 10,
      total: 4,
    },
  },
}
`;

snapshot[`search manga 1`] = `
{
  Page: {
    media: [
      {
        id: 102988,
        title: {
          english: "Tokyo Revengers",
          romaji: "Tokyo卍Revengers",
        },
      },
      {
        id: 114523,
        title: {
          english: null,
          romaji: "Baki Gaiden: Revenge Tokyo",
        },
      },
      {
        id: 152895,
        title: {
          english: null,
          romaji: "Tokyo卍Revengers Full Color Tanpenshuu",
        },
      },
      {
        id: 151997,
        title: {
          english: null,
          romaji: "Tokyo卍Revengers: Baji Keisuke Kara no Tegami ",
        },
      },
    ],
    pageInfo: {
      currentPage: 1,
      hasNextPage: false,
      lastPage: 1,
      perPage: 10,
      total: 4,
    },
  },
}
`;
